flow/project.py

```diff
@@ -3927,75 +3929,18 @@ def labels(self, job):
-
-        Any FlowOperation is associated with a specific command, which should
-        be a function of :class:`~signac.contrib.job.Job`. The command (cmd)
-        can be stated as function, either by using str-substitution based on a
-        job's attributes, or by providing a unary callable, which expects an
-        instance of job as its first and only positional argument.
-
-        For example, if we wanted to define a command for a program called
-        'hello', which expects a job id as its first argument, we could
-        construct the following two equivalent operations:
-
-        .. code-block:: python
-
-            op = FlowOperation('hello', cmd='hello {job.id}')
-            op = FlowOperation('hello', cmd=lambda 'hello {}'.format(job.id))
-
-        Here are some more useful examples for str-substitutions:
-
-        .. code-block:: python
-
-            # Substitute job state point parameters:
-            op = FlowOperation('hello', cmd='cd {job.ws}; hello {job.sp.a}')
-
-        Preconditions (pre) and postconditions (post) can be used to trigger an
-        operation only when certain conditions are met. Conditions are unary
-        callables, which expect an instance of job as their first and only
-        positional argument and return either True or False.
-
-        An operation is considered "eligible" for execution when all
-        preconditions are met and when at least one of the postconditions is
-        not met. Preconditions are always met when the list of preconditions
-        is empty. Postconditions are never met when the list of postconditions
-        is empty.
-
-        Please note, eligibility in this contexts refers only to the workflow
-        pipeline and not to other contributing factors, such as whether the
-        job-operation is currently running or queued.
-
-        Parameters
-        ----------
-        name : str
-            A unique identifier for this operation, which may be freely chosen.
-        cmd : str or callable
-            The command to execute operation; should be a function of job.
-        pre : sequence of callables
-            List of preconditions. (Default value = None)
-        post : sequence of callables
-            List of postconditions. (Default value = None)
-        \*\*kwargs
-            Keyword arguments passed as directives.
-
-        """
-        if name in self.operations:
-            raise KeyError("An operation with this identifier is already added.")
-        op = self.operations[name] = FlowCmdOperation(cmd=cmd, pre=pre, post=post)
-        if name in self._groups:
-            raise KeyError("A group with this identifier already exists.")
-        self._groups[name] = FlowGroup(
-            name, operations={name: op}, operation_directives=dict(name=kwargs)
+    @deprecated(
+        deprecated_in="0.14",
+        removed_in="0.16",
+        current_version=__version__,
+        details="Method has been removed.",
+    )
```

Comment:
Does it make sense to use the `@deprecated` decorator here since the function was not deprecated, but just removed?

Reply:
@csadorf The function hasn't been removed -- its _implementation_ has been removed. This approach allows us to raise an `AttributeError` with a helpful message, until we remove the function entirely. The deprecation warning is possibly redundant because that user message contains the same info, but I thought it was better to include because developers might search the codebase for `deprecated` when finding code to remove around release time.